plane/src/typed_unix_socket/mod.rs

```diff
@@ -0,0 +1,299 @@
+
+        assert_eq!(response, "Hello, client!");
+    }
+
+    #[tokio::test]
+    async fn test_client_to_server_ad_hoc() {
+        let socket_path = create_temp_socket_path();
+
+        let server = TypedUnixSocketServer::<String, String>::new(socket_path.to_str().unwrap())
+            .await
+            .unwrap();
+        let client = TypedUnixSocketClient::<String, String>::new(socket_path.to_str().unwrap())
+            .await
+            .unwrap();
+
+        // Subscribe to server events
+        let mut event_rx = server.subscribe_events();
+
+        // Send an ad-hoc message from the client
+        let ad_hoc_message = "Ad-hoc message from client".to_string();
+        client.send_message(ad_hoc_message.clone()).await.unwrap();
+
+        // Receive the ad-hoc message on the server
+        let received_message = event_rx.recv().await.unwrap();
+        assert_eq!(received_message, ad_hoc_message);
+    }
+
+    #[tokio::test]
+    async fn test_server_to_client_ad_hoc() {
+        let socket_path = create_temp_socket_path();
+
+        let server = TypedUnixSocketServer::<String, String>::new(socket_path.to_str().unwrap())
+            .await
+            .unwrap();
+        let client = TypedUnixSocketClient::<String, String>::new(socket_path.to_str().unwrap())
+            .await
+            .unwrap();
+
+        // Subscribe to client events
+        let mut event_rx = client.subscribe_events();
+
+        // Send an ad-hoc message from the server
+        let ad_hoc_message = "Ad-hoc message from server".to_string();
+        server.send_message(ad_hoc_message.clone()).await.unwrap();
+
+        // Receive the ad-hoc message on the client
+        let received_message = event_rx.recv().await.unwrap();
+        assert_eq!(received_message, ad_hoc_message);
+    }
+
+    #[tokio::test]
+    async fn test_concurrent_requests_responses() {
+        let socket_path = create_temp_socket_path();
+
+        let server = TypedUnixSocketServer::<String, String>::new(&socket_path)
+            .await
+            .unwrap();
+        let client = TypedUnixSocketClient::<String, String>::new(&socket_path)
+            .await
+            .unwrap();
+
+        // Spawn a task to handle server requests
+        spawn(async move {
+            let mut request_rx = server.subscribe_requests();
+            while let Ok(request) = request_rx.recv().await {
+                let response = format!("Response to {}", request.message);
+                server.send_response(&request, response).await.unwrap();
+            }
+        });
+
+        let mut handles = vec![];
+        for i in 0..10 {
+            let client = client.clone();
+            handles.push(spawn(async move {
+                let request = format!("Request {}", i);
+                let response = client.send_request(request).await.unwrap();
+                assert_eq!(response, format!("Response to Request {}", i));
+            }));
+        }
```

Comment:
I don't think this is a meaningfully different test from this:

```rust
for i in 0..10 {
    let request = format!("Request {}", i);
    let response = client.send_request(request).await.unwrap();
    assert_eq!(response, format!("Response to Request {}", i));
}
```

_Technically_, the runtime is allowed (as a matter of what it guarantees) to poll each task in arbitrary order, but it won't do that in practice, which means that we aren't actually testing any concurrency with this, and it could give us a false sense of confidence.

If you want to test that everything works if responses come out of order, consider having the server sleep, i.e something like this:

```rust
spawn(async move {
    let mut request_rx = server.subscribe_requests();
    while let Ok(request) = request_rx.recv().await {
        let req_num: i32 = request.parse().unwrap();
        tokio::sleep(Duration::from_millis(req_num * 100).await;
        let response = format!("Response to {}", request.message);
        server.send_response(&request, response).await.unwrap();
    }
});

let mut handles = vec![];
for i in 0..10 {
    let request = format!("{}", i);
    handles.push(async move {
        client.send_request(request).await.unwrap();
    });
}

let result = futures::future::join_all(handles);

for i in 0..10 {
    assert_eq!(result[i], format!("Response to Request {}", i));
}
```

Reply:
I tried this out in #748. If it LGTY, let's do the same for other concurrency tests, and then we are good to merge.

Reply:
Awesome, just merged that in and I'll take care of adding this for the rest -- thanks for catching this!